homeassistant/components/notify/apns.py

```diff
@@ -0,0 +1,238 @@
+    return service
+
+
+class ApnsDevice(object):
+    """
+    Apns Device class.
+
+    Stores information about a device that is
+    registered for push notifications.
+    """
+
+    def __init__(self, push_id, name, tracking_device_id=None):
+        """Initialize Apns Device."""
+        self.device_push_id = push_id
+        self.device_name = name
+        self.tracking_id = tracking_device_id
+
+    @property
+    def push_id(self):
+        """The apns id for the device."""
+        return self.device_push_id
+
+    @property
+    def name(self):
+        """The friendly name for the device."""
+        return self.device_name
+
+    @property
+    def tracking_device_id(self):
+        """
+        Device Id.
+
+        The id of a device that is tracked by the device
+        tracking component.
+        """
+        return self.tracking_id
+
+    @property
+    def full_tracking_device_id(self):
+        """
+        Fully qualified device id.
+
+        The full id of a device that is tracked by the device
+        tracking component.
+        """
+        return DEVICE_TRACKER_DOMAIN + '.' + self.tracking_id
+
+    def __eq__(self, other):
+        """Return the comparision."""
+        if isinstance(other, self.__class__):
+            return self.push_id == other.push_id and self.name == other.name
+        return NotImplemented
+
+    def __ne__(self, other):
+        """Return the comparision."""
+        return not self.__eq__(other)
+
+
+class ApnsNotificationService(BaseNotificationService):
+    """Implement the notification service for the AWS SNS service."""
+
+    def __init__(self, hass, app_name, sandbox, cert_file):
+        """Initialize APNS application."""
+        self.hass = hass
+        self.app_name = app_name
+        self.sandbox = sandbox
+        self.certificate = cert_file
+        self.yaml_path = hass.config.path(app_name + '_' + APNS_DEVICES)
+        self.devices = {}
+        self.device_states = {}
+        if os.path.isfile(self.yaml_path):
+            self.devices = {
+                str(key): ApnsDevice(
+                    str(key),
+                    value.get('name'),
+                    value.get('tracking_device_id')
+                )
+                for (key, value) in
+                load_yaml_config_file(self.yaml_path).items()
+            }
+
+        def state_changed_listener(entity_id, from_s, to_s):
+            """
+            Listener for sate change.
+
+            Track device state change if a device
+            has a tracking id specified.
+            """
+            self.device_states[entity_id] = str(to_s.state)
+            return
+
+        tracking_ids = [
+            device.full_tracking_device_id
+            for (key, device) in self.devices.items()
+            if device.tracking_device_id is not None
+        ]
+        track_state_change(hass, tracking_ids, state_changed_listener)
+
+    @staticmethod
+    def write_device(out, device):
```

Comment:
This writes out YAML, its a [flow mapping](http://www.yaml.org/spec/1.2/spec.html#id2790832) instead of the usual block mapping. It means that if the device name isn't supplied I can have an empty mapping for example:

`123456: {}`

Or if a name is supplied it will look like this:

`123456: { name: Test Phone }`